algorithms/src/Kokkos_RadixSort.hpp

```diff
@@ -0,0 +1,363 @@
+      }
+
+      // Number of bits is always even, and we know on odd numbered
+      // iterations we are reading from m_key_scratch and writing to keys
+      // So when this loop ends, keys will contain the results
+    }
+  }
+
+  // Directly re-arrange the entries of keys, optionally storing the permutation
+  template <bool store_permutation = false, typename U, typename ExecutionSpace>
+  void sortByKeys(ExecutionSpace const& exec, View<KeyType*, MemorySpace> keys,
+                  View<U*, MemorySpace> values) {
+    // Almost identical to create_indirection_array, except actually permute the
+    // input
+    const auto n = keys.extent(0);
+    RangePolicy<ExecutionSpace, Kokkos::IndexType<IndexType>> policy(exec, 0,
+                                                                     n);
+
+    if constexpr (store_permutation) {
+      Kokkos::parallel_for(
+          "Kokkos::RadixSorter sortByKeys store_permutation", policy,
+          KOKKOS_LAMBDA(IndexType i) { m_index_old(i) = i; });
+    }
+
+    auto values_scratch = View<U*, MemorySpace>(
+        view_alloc(exec, "Kokkos::RadixSorter radix_sorter_values_scratch",
+                   Kokkos::WithoutInitializing),
+        n);
+
+    using KeyFunctor = decltype(KeyFromView{keys});
+
+    (void)values;  // Avoid a false warning from icpc about values being unused
+
+    for (int i = 0; i < KeyFunctor::num_bits; ++i) {
+      KeyFunctor key_functor = KeyFromView{keys};
+
+      step(
+          policy, key_functor, i, KOKKOS_LAMBDA(size_t id) { return id; });
+      if constexpr (store_permutation) {
+        permute_by_scan<KeyType, U, IndexType>(policy, {m_key_scratch, keys},
+                                               {values_scratch, values},
+                                               {m_index_new, m_index_old});
+      } else {
+        permute_by_scan<KeyType, U>(policy, {m_key_scratch, keys},
+                                    {values_scratch, values});
+      }
+
+      // Number of bits is always even, and we know on odd numbered
+      // iterations we are reading from m_key_scratch/values_scratch and writing
+      // to keys/values. So, when this loop ends, keys/values will contain the
+      // results
+    }
+  }
+
+  template <typename ExecutionSpace>
+  void apply_permutation(ExecutionSpace const& exec,
+                         View<KeyType*, MemorySpace> v) {
+    parallel_for(
+        "Kokkos::RadixSorter apply_permutation",
+        RangePolicy<ExecutionSpace, Kokkos::IndexType<IndexType>>(exec, 0,
+                                                                  v.extent(0)),
+        KOKKOS_LAMBDA(IndexType i) { m_key_scratch(i) = v(m_index_old(i)); });
+    deep_copy(exec, v, m_key_scratch);
+  }
+
+ private:
+  template <typename... U, typename Policy>
+  void permute_by_scan(
+      Policy policy,
+      Kokkos::pair<View<U*, MemorySpace>&, View<U*, MemorySpace>&>... views) {
```

Comment:
The pair of references here is dangerous with the HPX backend. The swap at the end of the function may end up swapping the views before the `parallel_for` has operated on the views. The views should be captured by value.